manager/scheduler/constraint.go

```diff
@@ -44,26 +44,36 @@ func (f *ConstraintFilter) Check(n *NodeInfo) bool {
 			if !constraint.Match(n.Description.Hostname) {
 				return false
 			}
+		case "node.role":
+			if !constraint.Match(n.Spec.Role.String()) {
+				return false
+			}
 		default:
 			// default is node label in form like 'node.labels.key==value'
 			// if it is not well formed, always fails it
 			if !strings.HasPrefix(constraint.Key, "node.labels.") {
 				return false
 			}
-			// if the node doesn't have any label,
-			// it's equivalent to match an empty value.
-			// that is, 'node.labels.key!=value' should pass and
-			// 'node.labels.key==value' should fail
-			if n.Spec.Annotations.Labels == nil {
-				if !constraint.Match("") {
-					return false
+
+			// node has labels from node.Spec.Annotations.Labels and node.Description.Engine.Labels
+			// if a label exists on both, node.Spec.Annotations.Labels has precedence
+			combinedLabels := make(map[string]string)
+			if n.Description != nil && n.Description.Engine != nil && n.Description.Engine.Labels != nil {
```

Comment:
I really want the use-case of creating an overlay of labels that allows me to have different security levels for hosts (low, medium, high), and this *has* to be enforced manager side, and can't be changed by the workers.

I think we should support this use-case. Also, @dongluochen, do you mind auditing the code to make sure we're not trusting what the workers say, ensuring that this is actually dependable?